src/pip/_internal/vcs/git.py

```diff
@@ -194,10 +194,15 @@ def resolve_revision(cls, dest, url, rev_options):
                 rev,
             )
 
-        if not rev.startswith('refs/'):
+        if not (
+            rev.startswith('refs/')
+            or (looks_like_hash(rev) and not cls.has_commit(dest, rev))
+        ):
```

Comment:
My head spins trying to parse this 🤯 

Maybe…

```suggestion
        if not rev.startswith('refs/')
            and (not looks_like_hash(rev) or cls.has_commit(dest, rev)):
```

Or extract this into a function?

```python
def _seems_available(rev):
    if rev.startswith('refs/'):
        return False
    if not looks_like_hash(rev):
        return True
    if cls.has_commit(dest, rev):
        return True
    return False
```

Reply:
+1 for extracting into a function.